internal/vulnstore/postgres/gc.go

```diff
@@ -0,0 +1,278 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+
+	"github.com/google/uuid"
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"golang.org/x/sync/semaphore"
+)
+
+// GC is split into two phases, first it will identify any update operations
+// which are older then the provided keep value and delete these.
+//
+// Next it will perform chunked deletions of any vulns from the vuln table
+// which are not longer referenced by update operations.
+//
+// The GC is throttled to not overload the database with cascade deletes.
+// If a full GC is required run this method until the returned int64 value
+// is 0.
+func (s *Store) GC(ctx context.Context, keep int) (int64, error) {
+	const (
+		// GCThrottle sets a limit for the number of deleted update operations
+		// (and subsequent cascade deletes in the uo_vuln table) that can occur in a GC run.
+		GCThrottle = 50
+	)
+
+	// obtain update operations which need deletin'
+	ops, totalOps, err := eligibleUpdateOpts(ctx, s.pool, keep)
+	if err != nil {
+		return 0, err
+	}
+
+	// delete em', but not too many...
+	if totalOps >= GCThrottle {
+		ops = ops[:GCThrottle]
+	}
+
+	deletedOps, err := s.DeleteUpdateOperations(ctx, ops...)
+	if err != nil {
+		return totalOps - deletedOps, err
+	}
+
+	// get all updaters we know about.
+	updaters, err := distinctUpdaters(ctx, s.pool)
+	if err != nil {
+		return totalOps - deletedOps, nil
```

Comment:
So if it was unable to get the updaters, the error is just swallowed?

Reply:
nope that's a bug.